modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TypeUtils.java

```diff
@@ -138,11 +138,10 @@ public static boolean needCast(RelDataTypeFactory factory, RelDataType fromType,
         // No need to cast if the source type precedence list
         // contains target type. i.e. do not cast from
         // tinyint to int or int to bigint.
-        if (fromType.getPrecedenceList().containsType(toType)
-            && SqlTypeUtil.isIntType(fromType)
-            && SqlTypeUtil.isIntType(toType)) {
+        // Currently, RelDataTypeFactoryImpl#CLASS_FAMILIES doesn't consider the byte type.
+        if (SqlTypeUtil.isIntType(fromType) && SqlTypeUtil.isIntType(toType) && (fromType.getPrecedenceList().containsType(toType))
+            || fromType.getSqlTypeName() == SqlTypeName.TINYINT || toType.getSqlTypeName() == SqlTypeName.TINYINT)
```

Comment:
`&&` has higher priority than `||`, parentheses are required. 
In your case any call (for example with `VARCHAR` type) will return `false` if `fromType` or `toType` will be `TINYINT`.
Also, I don't understand why we refer to `RelDataTypeFactoryImpl#CLASS_FAMILIES`. 
`getPrecedenceList()` returns `SqlTypeExplicitPrecedenceList` for integer type, which contains `TINYINT` as well. I've commented this change and test still working. Do we really need this change?

Reply:
Without this change `StatisticsPlannerTest#testBorders()` will fail on the `Byte` table field. Rewrote.